reggie/ingestion/preprocessor/west_virginia_preprocessor.py

```diff
@@ -1,87 +1,361 @@
+            "unaffiliated": re.compile("(unaffiliated|no.*affiliation)"),
+        }
+
+        for k, v in lookup.items():
+            if v.match(value):
+                return k
+
+        return "unknown"
+
+    def clean_challenged_value(self, value):
+        true_regex = re.compile("^y$", re.IGNORECASE)
+        false_regex = re.compile("^n$", re.IGNORECASE)
+
+        if not isinstance(value, str):
+            return None
+        elif true_regex.match(value):
+            return True
+        elif false_regex.match(value):
+            return False
+
+        return None
+
+    def derive_votetype(self, row):
+        """
+        Given row in voter history, make a consistent vote type.
+        """
+        true_regex = re.compile("^y$", re.IGNORECASE)
+        # false_regex = re.compile("^n$", re.IGNORECASE)
+
+        if isinstance(row["fl_absentee"], str) and true_regex.match(row["fl_absentee"]):
+            return "absentee"
+        elif isinstance(row["fl_early_voting"], str) and true_regex.match(
+            row["fl_early_voting"]
+        ):
+            return "early"
+        elif isinstance(row["fl_regular"], str) and true_regex.match(row["fl_regular"]):
+            return "regular"
+
+        return "unknown"
+
+    def config_lookup_to_dict(self, config_values):
+        """
+        Takes a config like:
+        gender_codes:
+            Male:
+                - M
+                - m
+            Female:
+                - F
+                - f
+            Unknown:
+                - N
+                - nan
+                - ""
+
+        And turns it into:
+            { "M": "Male", "m": "Male" ...}
+        """
+        converted = {}
+        for c, v in config_values.items():
+            for i in v:
+                if i is None:
+                    converted[""] = c
+                elif i == "nan":
+                    # converted[np.nan] = c
+                    converted[float("nan")] = c
+                    converted["nan"] = c
+                else:
+                    converted[i] = c
+
+        return converted
+
+    def config_lookup_to_valid_list(self, config_values):
+        """
+        Takes a config like:
+        gender_codes:
+            Male:
+                - M
+                - m
+            Female:
+                - F
+                - f
+            Unknown:
+                - N
+                - nan
+                - ""
+
+        And turns it into:
+            ["Male", "Female", "Unknown"]
+        """
+        return config_values.keys()
+
+    def check_column_has_valid_values(self, df, column, valid_values):
+        config = self.config.data
+        df_invalid_values = df[~df[column].isin(valid_values)]
+        if len(df_invalid_values) > 0:
+            raise InvalidDataError(
+                f"{config['state']} found the following invalid values in the '{column}' column': {df_invalid_values[column].unique()}"
+            )
```

Comment:
(Copied from the same comment on Nick's PR) - One thing you're missing in the preprocessor is 2 of Tommi's newish pipeline checks: 1) "file_check" which needs you to define "expected_number_of_files" in the yaml, and 2) "column_check" which validates the columns you receive in the voter file. You can see examples of how those 2 pipeline checks are done in other states.